Source/WebCore/platform/LibWebRTCVPXVideoEncoder.cpp

```diff
@@ -0,0 +1,180 @@
+
+    void postTask(Function<void()>&& task) { m_postTaskCallback(WTFMove(task)); }
+    void encode(VideoEncoder::RawFrame&&, bool shouldGenerateKeyFrame, VideoEncoder::EncodeCallback&&);
+    void close() { m_isClosed = true; }
+private:
+    LibWebRTCVPXInternalVideoEncoder(LibWebRTCVPXVideoEncoder::Type, const VideoEncoder::Config&, VideoEncoder::OutputCallback&&, VideoEncoder::PostTaskCallback&&);
+    webrtc::EncodedImageCallback::Result OnEncodedImage(const webrtc::EncodedImage&, const webrtc::CodecSpecificInfo*) final;
+    void OnDroppedFrame(DropReason) final;
+
+    VideoEncoder::OutputCallback m_outputCallback;
+    VideoEncoder::PostTaskCallback m_postTaskCallback;
+    UniqueRef<webrtc::VideoEncoder> m_internalEncoder;
+    int64_t m_timestamp { 0 };
+    std::optional<uint64_t> m_duration;
+    bool m_isClosed { false };
+};
+
+LibWebRTCVPXVideoEncoder::LibWebRTCVPXVideoEncoder(Type type, const VideoEncoder::Config& config, OutputCallback&& outputCallback, PostTaskCallback&& postTaskCallback)
+    : m_internalEncoder(LibWebRTCVPXInternalVideoEncoder::create(type, config, WTFMove(outputCallback), WTFMove(postTaskCallback)))
+{
+}
+
+LibWebRTCVPXVideoEncoder::~LibWebRTCVPXVideoEncoder()
+{
+}
+
+void LibWebRTCVPXVideoEncoder::encode(RawFrame&& frame, bool shouldGenerateKeyFrame, EncodeCallback&& callback)
+{
+    vpxQueue().dispatch([frame = WTFMove(frame), shouldGenerateKeyFrame, encoder = m_internalEncoder, callback = WTFMove(callback)]() mutable {
+        encoder->encode(WTFMove(frame), shouldGenerateKeyFrame, WTFMove(callback));
+    });
+}
+
+void LibWebRTCVPXVideoEncoder::flush(Function<void()>&& callback)
+{
+    vpxQueue().dispatch([encoder = m_internalEncoder, callback = WTFMove(callback)]() mutable {
+        encoder->postTask(WTFMove(callback));
+    });
+}
+
+void LibWebRTCVPXVideoEncoder::reset()
+{
+    m_internalEncoder->close();
+}
+
+void LibWebRTCVPXVideoEncoder::close()
+{
+    m_internalEncoder->close();
+}
+
+
+void LibWebRTCVPXInternalVideoEncoder::encode(VideoEncoder::RawFrame&& rawFrame, bool shouldGenerateKeyFrame, VideoEncoder::EncodeCallback&& callback)
+{
+    m_timestamp = rawFrame.timestamp;
+    m_duration = rawFrame.duration;
+
+    auto frameType = shouldGenerateKeyFrame ? webrtc::VideoFrameType::kVideoFrameKey : webrtc::VideoFrameType::kVideoFrameDelta;
+    std::vector<webrtc::VideoFrameType> frameTypes { frameType };
+
+    auto frameBuffer = webrtc::pixelBufferToFrame(rawFrame.frame->pixelBuffer());
+
+    webrtc::VideoFrame frame { frameBuffer, webrtc::kVideoRotation_0, rawFrame.timestamp };
+    auto error = m_internalEncoder->Encode(frame, &frameTypes);
+
+    m_postTaskCallback([protectedThis = Ref { *this }, error, callback = WTFMove(callback)]() mutable {
+        if (protectedThis->m_isClosed)
+            return;
+
+        String result;
+        if (error)
+            result = makeString("VPx encoding failed with error ", error);
+        callback(WTFMove(result));
+    });
```

Comment:
It seems like it would be useful to have console logging throughout the WebCodecs code